planner/core/planbuilder.go

```diff
@@ -1222,49 +1222,44 @@ func removeTiflashDuringStaleRead(paths []*util.AccessPath) []*util.AccessPath {
 
 func (b *PlanBuilder) buildSelectLock(src LogicalPlan, lock *ast.SelectLockInfo) (*LogicalLock, error) {
 	selectLock := LogicalLock{
-		Lock:             lock,
-		tblID2Handle:     b.handleHelper.tailMap(),
-		partitionedTable: b.partitionedTable,
+		Lock:         lock,
+		tblID2Handle: b.handleHelper.tailMap(),
 	}.Init(b.ctx)
 	selectLock.SetChildren(src)
 
 	if len(b.partitionedTable) > 0 {
-		// If a chunk row is read from a partitioned table, which partition the row
-		// comes from is unknown. With the existence of Join, the situation could be
-		// even worse: SelectLock have to know the `pid` to construct the lock key.
-		// To solve the problem, an extra `pid` column is add to the schema, and the
-		// DataSource need to return the `pid` information in the chunk row.
-		err := addExtraPIDColumnToDataSource(src, &selectLock.extraPIDInfo)
-		if err != nil {
-			return nil, err
-		}
+		tables := make([]table.PartitionedTable, 0, len(b.partitionedTable))
+		tables = collectPartitionTable(selectLock, tables)
+
 		// TODO: Dynamic partition mode does not support adding extra pid column to the data source.
 		// (Because one table reader can read from multiple partitions, which partition a chunk row comes from is unknown)
 		// So we have to use the old "rewrite to union" way here, set `flagPartitionProcessor` flag for that.
 		b.optFlag = b.optFlag | flagPartitionProcessor
+		selectLock.partitionedTable = tables
 	}
 	return selectLock, nil
 }
 
-func addExtraPIDColumnToDataSource(p LogicalPlan, info *extraPIDInfo) error {
+func collectPartitionTable(p LogicalPlan, input []table.PartitionedTable) []table.PartitionedTable {
 	switch raw := p.(type) {
 	case *DataSource:
 		// Fix issue 26250, do not add extra pid column to normal table.
 		if raw.tableInfo.GetPartitionInfo() == nil {
-			return nil
-		}
-		raw.addExtraPIDColumn(info)
-		return nil
+			return input
+		}
+		input = append(input, raw.table.(table.PartitionedTable))
+	case *LogicalApply:
+		// For logical apply, the query should only lock t1, although it's converted to join:
+		//     select * from t1 where t1.id in (select id from t2) for update
+		//
+		// See issue https://github.com/pingcap/tidb/issues/30382
+		input = collectPartitionTable(raw.Children()[0], input)
 	default:
-		var err error
 		for _, child := range p.Children() {
-			err = addExtraPIDColumnToDataSource(child, info)
-			if err != nil {
-				return err
-			}
+			input = collectPartitionTable(child, input)
```

Comment:
Why do you still collect the partition table of the aggregation?

Reply:
And also, you don't need to collect the things after a subquery.
For the two following queries
```sql
select * from t, (select * from t1) t1 where t.c=t1.id for update;
select * from t, t1 where t.c=t1.id for update;
```
The first would not lock the t1, only the second one would.

Maybe you can refer to the `handleHelper` of the `PlanBuilder`
```go
	// handleHelper records the handle column position for tables. Delete/Update/SelectLock/UnionScan may need this information.
	// It collects the information by the following procedure:
	//   Since we build the plan tree from bottom to top, we maintain a stack to record the current handle information.
	//   If it's a dataSource/tableDual node, we create a new map.
	//   If it's an aggregation, we pop the map and push a nil map since no handle information left.
	//   If it's a union, we pop all children's and push a nil map.
	//   If it's a join, we pop its children out then merge them and push the new map to stack.
	//   If we meet a subquery, it's clear that it's an independent problem so we just pop one map out when we finish building the subquery.
	handleHelper *handleColHelper
```